includes/spam.php

```diff
@@ -0,0 +1,139 @@
+<?php
+/**
+ * Code related to spam detection and prevention.
+ */
+// Constants. Define these in wp-config.php to override.
+if ( ! defined( 'PMPRO_SPAM_ACTION_NUM_LIMIT' ) ) {
+    define( 'PMPRO_SPAM_ACTION_NUM_LIMIT', 10 );
+}
+if ( ! defined( 'PMPRO_SPAM_ACTION_TIME_LIMIT' ) ) {
+    define( 'PMPRO_SPAM_ACTION_TIME_LIMIT', 900 );  // in seconds
+}
+
+/**
+ * Is the current visitor a spammer?
+ * @since 2.7
+ */
+function pmpro_is_spammer() {
+    $is_spammer = false;
+    
+    $activity = pmpro_get_spam_activity();
+    if ( count( $activity ) >= PMPRO_SPAM_ACTION_NUM_LIMIT ) {
+        $is_spammer = true;
+    }
+    
+    return apply_filters( 'pmpro_is_spammer', $is_spammer, $activity );
+}
+
+/**
+ * Get spam activity by IP.
+ * @since 2.7
+ * @param string $ip    The IP address to get activity for.
+ * @return array|False  The array of activity if successful, false if no IP. 
+ */
+function pmpro_get_spam_activity( $ip = null ) {
+    if ( empty( $ip ) ) {
+        $ip = pmpro_get_ip();        
+    }
+    
+    // If we can't determine the IP, let's bail.
+    if ( empty( $ip ) ) {
+        return false;
+    }
+    
+    $transient_key = 'pmpro_spam_activity_' . $ip;

```

Comment:
If $ip is provided, should we sanitize/clean `$ip` in any way before getting transient?

Reply:
No. Probably happened when copying some of this code from a different branch on GitHub or something. I like tabs 4 spaces long. I can lint when we're all done.